### ISSUE TITLE
investigate and likely switch to openvla-oft

### ISSUE THREAD

**Issue by mehhl** [nomagiclab/openvla#4]
there is a new version of OpenVLA, called OpenVLA-OFT [link](https://openvla-oft.github.io) which claims SOTA (as in, better than pi0) performance on variousb enchmarks thanks to enhanced finetuning. The enhancements go far enough to significantly change the model architecture (up to 3 video streams as input, proprio as input, continuous actions rather than tokenization, ...) and the authors are 3 of the original OpenVLA paper authors, one of whom is also a co-founder of Physical Intelligence. It would seem prudent to follow this program:

1. See if OpenVLA-OFT gives some meaningful improvement on our own tasks (nomagic-simple-box, etc.) 
2. If the answer to 1. is anything like "yes", switch any/all OpenVLA experiments to OpenVLA-OFT asap.

**Comment by mehhl** [comment]
For now finetuning our first model ([here it is](https://wandb.ai/robotgeneralist/ur5e/runs/facdgflc?nw=nwuserelmehlan)), we'll see in the morning how it behaves. In this one I've basically turned on _all_ the features of OpenVLA-OFT except diffusion, just wanted to see what happens. For now, the loss drops, but with an amplitude reminescent of the very worst parts of reinforcement learning:

![Image](https://github.com/user-attachments/assets/0193d4d7-94c0-4ea8-b1f1-e55fc3e77f9f)

Next I should train a sequence (a gradient, if you will) of models from completely vanilla OpenVLA, no FiLM, no nothing but incorporating [recent changes](https://github.com/nomagiclab/openvla/pull/5) to the fine-tuning code, all the way to full OpenVLA-OFT+. Mostly just to see if those changes in finetuning code degrade performance due to bugs. First goal is to reproduce vanilla OpenVLA's best performance with new code.

It also now becomes relevant to implement [inference with OpenVLA](https://github.com/orgs/nomagiclab/projects/6) asap-ish.

**Comment by mehhl** [comment]
I mean, I don't know. Replay on episode 0:

![Image](https://github.com/user-attachments/assets/6556ec9a-643f-4378-bfc0-0d9aa5c1ad68)

![Image](https://github.com/user-attachments/assets/a4fb5809-8dae-483a-bb23-e89f4597e686)

![Image](https://github.com/user-attachments/assets/067d9a89-3555-48c6-b692-16c23d65a129)

![Image](https://github.com/user-attachments/assets/e636eeb0-9501-4c8f-b3bd-d0a8fd81de25)

![Image](https://github.com/user-attachments/assets/7cd80a27-b187-4adb-a7ea-2e0effe4574e)

![Image](https://github.com/user-attachments/assets/c57053cb-24ac-4a3c-a6bf-c374f480ad6e)

![Image](https://github.com/user-attachments/assets/52a6a9c1-6ae4-406a-84cb-24a41e4a39f6)

It is certainly something rather than nothing, but also the model seems to lack courage in some sense, and exhibit frankly bizarre biases (such as on dimension 1). If I had to sum it up in one sentence, the model moves well in some directions but not in others.

I've started a huge rollout of the model over all 230 episodes in the finetuning dataset, it will take about 10 hours. I'd be happy to find out what is the source of model's apparent bias in when to follow demonstrations faithfully vs when to not.

**Comment by mehhl** [comment]
Another idea is that the dataset might be just a bit too large (230 episodes, of which something like 210 are used for training and 20 for validation). Maybe we should either finetune for longer or on a smaller subset.

**Comment by mehhl** [comment]
here's another fun visualization:

![Image](https://github.com/user-attachments/assets/e8c3d3e8-2af0-490a-831a-97bdafceaee3)

![Image](https://github.com/user-attachments/assets/071fb650-744b-468d-901c-878bc89445f9)

![Image](https://github.com/user-attachments/assets/bf610464-d9ff-45ae-9fb6-774393ffaa2c)

It plots absolute magnitude of error in the fitted actions $\lvert \hat{a}\_{(\cdot)} - a\_{(\cdot)}\rvert$, $(\cdot) \in \{x, y, z\}$ as function of $XYZ$ displacement of the groundtruth action, ie $(a_x, a_y, a_z)$. Here's yet another fun one:

![Image](https://github.com/user-attachments/assets/b3f4be0b-6b6c-4133-918a-134984c1803c)

![Image](https://github.com/user-attachments/assets/3493ec01-a8fe-4745-a06b-b84fe7ed6023)

![Image](https://github.com/user-attachments/assets/8cf326a2-1276-40e8-8034-fe9f069f9be0)

This one plots $\lvert \hat{a}\_{(\cdot)} - a\_{(\cdot)}\rvert$ against $\lVert (a_x, a_y, a_z) \rVert$. 

Here are predicted vs gt acitons for the first 20 episodes:

![Image](https://github.com/user-attachments/assets/64aabb8f-c9a8-4035-bab2-4c994a28d2c2)

![Image](https://github.com/user-attachments/assets/11b5dced-f4ae-4ed9-a7d9-de85e61d5099)

![Image](https://github.com/user-attachments/assets/d0dd40c9-578f-41cf-8ca5-c68ee6922c10)

![Image](https://github.com/user-attachments/assets/d23ecac8-2dc0-4e2f-b5b6-c9ddad371685)

![Image](https://github.com/user-attachments/assets/521f4bd7-fe3a-42a2-957a-686f2dad6008)

![Image](https://github.com/user-attachments/assets/3a272513-dff0-4bc1-befd-a360e03836d8)

![Image](https://github.com/user-attachments/assets/4eb57493-e430-4684-ab20-ff27ddc9bbc5)

our problems reside IMO chiefly in the dimension $Z$, or "dimension 2" above. If we were able to make model move confidently downwards, I think that would imply most other problems get solved as well. It is clear that the model has a systemic bias in all three cardinal directions:

![Image](https://github.com/user-attachments/assets/3d37496c-47b6-4215-ae12-b8126f18ee0d)

![Image](https://github.com/user-attachments/assets/c2ae1a7c-d4cf-49f5-a062-711bbe123269)

![Image](https://github.com/user-attachments/assets/272a11ea-3bd9-486b-b9e6-a5b7f315addb)

which I personally would blame on undertrained action head, then _possibly_ on some unnormalization quirk. 

I think the model is undertrained, because there is a certain _softness_ to its errors: they seem like artifacts of some neural process, as evidenced by the fact that the model tends to make similar mistakes in similar places. The artifacts are intricate at the exact scale (5-10 timesteps) I would expect if they came from an undertrained model. Longer clusters of movement (15 timesteps or more) are generally recognized and imitated by the model, but at 5-10 timesteps it didn't have the time to notice most of them yet.

I'll try two things next (in that order):

1. A new training run that records more evaluation info, so that we can see the situation depicted above develop over time. I'll also decrease number of train trajectories to 50 (training on 200 trajs with two A100s for 48h is just silly).

2. A much bigger training run, on the order of 4 A100s for 96h, or 8 A100s for 48h.
